File: Monte_Carlo/mc.cpp
#include <iostream>
#include <stdlib.h>
#include <math.h>
#include <fstream>
#include <string>

#define PI 3.14159265

using namespace std;

int main() {

    string str;
    int number_of_particles;
    int number_of_steps;
    double material_thickness;
    double travel_distance;

    ifstream indata;
    indata.open("input.txt");
    indata >> str >> number_of_particles;
    indata >> str >> number_of_steps;
    indata >> str >> material_thickness;
    indata >> str >> travel_distance;
    indata.close();


    int escaped_particles = 0;

    for (int i = 0; i < number_of_particles; i++) {
        double particle_position = 0.0;
        int current_step = 0;

        while (current_step < number_of_steps) {
            double theta = ((float) rand()/RAND_MAX)*2*PI;
            particle_position += travel_distance*cos(theta);

            if (particle_position < 0.0 || particle_position > material_thickness) {
                current_step = number_of_steps;
            } else {
                current_step += 1;
            }
        }

        if (particle_position >= material_thickness) {
            escaped_particles += 1;
        }
    }

    cout<<"Escaped Particles: "<<escaped_particles<<endl;
    cout<<"Percent Escaped: "<<double(escaped_particles)*100/double(number_of_particles)<<endl;

    return 0;
}

File: Monte_Carlo/mc.py
import random
import math

number_of_particles = 100
number_of_steps = 6
material_thickness = 15
travel_distance = 5
escaped_particles = 0

# Looping over the large number of particles
for i in range(number_of_particles):
    # Initializing the particle
    particle_position = 0
    current_step = 0
    # Looping over the steps
    while (current_step < number_of_steps):
        # Finding the bounce angle and moving the particle accordingly
        theta = random.random()*2*math.pi
        particle_position += travel_distance*math.cos(theta)
        # Exit conditions
        if (particle_position < 0 or particle_position > material_thickness):
            current_step = number_of_steps
        else:
            # Continue if exit condition wasn't met.
            current_step += 1
    
    # While loop ended, check if particle escaped.
    if particle_position >= material_thickness:
        escaped_particles += 1

# Getting final results
print("Escaped Particles: " + str(escaped_particles))
print("Percent Escaped: " + str(escaped_particles*100/number_of_particles)+"%")
